src/transformers/models/emu3/modular_emu3.py

```diff
@@ -0,0 +1,1982 @@
+            kernel_size=3,
+            stride=1,
+            padding=1,
+        )
+
+        temporal_down_blocks = int(math.log2(config.temporal_downsample_factor))
+        self.time_conv = nn.ModuleList()
+
+        for i in range(temporal_down_blocks):
+            conv = Emu3VQVAETemporalDownsample(out_channels, out_channels)
+            self.time_conv.append(conv)
+
+        self.time_res_stack = nn.Sequential(
+            *[
+                Emu3VQVAETemporalResnetBlock(
+                    in_channels=out_channels,
+                    out_channels=out_channels,
+                )
+                for _ in range(self.num_res_blocks)
+            ]
+        )
+
+    def forward(self, pixel_values: torch.LongTensor):
+        temporal_dim = pixel_values.shape[1]
+        pixel_values = pixel_values.reshape(-1, *pixel_values.shape[2:])
+
+        # downsampling
+        hidden_states = self.conv_in(pixel_values)
+        for i_level in range(self.num_resolutions):
+            for i_block in range(self.num_res_blocks):
+                hidden_states = self.down[i_level].block[i_block](
+                    hidden_states,
+                )
+                if len(self.down[i_level].attn) > 0:
+                    hidden_states = self.down[i_level].attn[i_block](hidden_states)
+            if i_level != self.num_resolutions - 1:
+                hidden_states = self.down[i_level].downsample(hidden_states)
+
+        # middle
+        hidden_states = self.mid.block_1(hidden_states)
+        hidden_states = self.mid.attn_1(hidden_states)
+        hidden_states = self.mid.block_2(hidden_states)
+
+        # end
+        hidden_states = self.norm_out(hidden_states)
+        hidden_states *= torch.sigmoid(hidden_states)
+        hidden_states = self.conv_out(hidden_states)
+
+        hidden_states = hidden_states.reshape(-1, temporal_dim, *hidden_states.shape[1:])
+        hidden_states = hidden_states.permute(0, 2, 1, 3, 4)
+
+        for conv in self.time_conv:
+            hidden_states = conv(hidden_states)
+            hidden_states *= torch.sigmoid(hidden_states)
+
+        hidden_states = self.time_res_stack(hidden_states)
+        hidden_states = hidden_states.permute(0, 2, 1, 3, 4)
+
+        return hidden_states
+
+
+class Emu3VQVAEDecoder(nn.Module):
+    def __init__(self, config: Emu3VQVAEConfig):
+        super().__init__()
+        self.base_channels = config.base_channels
+        self.num_resolutions = len(config.channel_multiplier)
+        self.num_res_blocks = config.num_res_blocks
+
+        quant_channels = config.embed_dim
+        block_in = config.base_channels * config.channel_multiplier[-1]
+        self.time_res_stack = nn.Sequential(
+            *[
+                Emu3VQVAETemporalResnetBlock(
+                    in_channels=config.latent_channels,
+                    out_channels=config.latent_channels,
+                )
+                for _ in range(config.num_res_blocks)
+            ]
+        )
```

Comment:
would prefer `nn.Modulist` 😅

Reply:
can do but that means we'll have a loop over each moduel to call forward, while in sequential it is cleaner to just call `forward` once on everything

Reply:
yep, but we'd rather have explicity as it's easier to debug